Return available data from stream, before waiting for more

LinusKardellInfobric opened the pull request:
Previously, Read and ReadAsync would only return once the output buffer had been filled, or the inner stream returned 0 (i.e. when the inner stream ended). That meant it would not return already decoded data until it got more data from the inner stream, even if no more data was currently available. Now, instead:

- If the previous call filled the output buffer (which means there could potentially still be data buffered in the decompression context), try flushing that out by running DecompressStream, before trying to read more input.
- As soon as we have some output available, return it, so that we won't stall on reading from the inner stream before returning the decompressed data.

This does have the drawback that each call to Read will only call innerStream.Read at most once, so the amount of data the caller can get from a single Read call is limited by inputBufferSize.

Fixes #32

oleg-st commented:
> inputBufferSize = bufferSize > 0 ? bufferSize : (int) Methods.ZSTD_CStreamInSize().EnsureZstdSuccess();

Looks like it should be `ZSTD_DStreamInSize` instead of `ZSTD_CStreamInSize`

LinusKardellInfobric commented:
> Looks like it should be `ZSTD_DStreamInSize` instead of `ZSTD_CStreamInSize`

Unrelated to this change, but OK, I changed that.

> Some tests have failed, such as `RoundTrip_BatchToStreaming`

Turns out when we're completely done with a frame, calling DecompressStream again with no input will return 9 (the minimum possible ZSTD frame size) rather than 0. I've changed it to only keep the result of DecompressStream calls that made some progress.
